### DISCUSSION TITLE
m-api Constructing the query string

### DISCUSSION THREAD

**Discussion by niro1987** [persandstrom/python-verisure#139] [Show and tell]
Hi,

I've been playing with the m-api brach. I've setup mitmproxy to intercept traffic from my mobile device.
While browsing through the calls made, I noticed a regularity. In testing, I found that you can construct your own query string.

For example, this query will fetch all installations and from those retrieve the `giid`, `alias` and `pinCodeLength`.
```graphql
query Installation($giid: String!) {
    installation(giid: $giid) {
        giid
        alias
        pinCodeLength
    }
}
```
replace the newline characters with `\n` te create the python string
```txt
"query Installation($giid: String!) {\n    installation(giid: $giid) {\n        giid\n        alias\n        pinCodeLength\n    }\n}"
```

Now, let's add the `armState`
```graphql
query Installation($giid: String!) {
    installation(giid: $giid) {
        giid
        alias
        pinCodeLength
        armState {
            statusType
            date
            name
            changedVia
        }
    }
}
```

as python string, this becomes:
```txt
"query Installation($giid: String!) {\n    installation(giid: $giid) {\n        giid\n        alias\n        pinCodeLength\n        armState {\n            statusType\n            date\n            name\n            changedVia\n        }\n    }\n}"
```

And why not add more;
```graphql
query Installation($giid: String!, $all: Boolean!) {
    installation(giid: $giid) {
        giid
        alias
        pinCodeLength
        broadband {
            testDate
            isBroadbandConnected
        }
        armState {
            statusType
            date
            name
            changedVia
        }
        doorWindows {
            device {
                deviceLabel
                area
            }
            state
            reportTime
        }
        smartLocks {
            lockStatus
            doorState
            lockMethod
            eventTime
            doorLockType
            device {
                deviceLabel
                area
            }
        }
        cameras(allCameras: $all) {
            device {
                deviceLabel
                area
            }
        }
        cameraContentProvider {
            latestImage {
                deviceLabel
                contentType
                contentUrl
                timestamp
            }
        }
    }
}
```

**Comment by persandstrom** [comment]
Great! The m-api is based on GraphQL, I have not dabbled in it much myself, but it seems quite easy to work with

**Comment by persandstrom** [comment]
My initial thought about not creating a big Overview function was that if one part of the request fails, the rest will still work. So for example if a field in "user_trackings" is changed, then "smartplugs" part will still work. If an overview function is created then everything will stop working at once if any field is changed. 

But looking at your overview function i believe that it might make more sense anyway, since it will create a much cleaner interface for home-assistant. If we go for an overview function I might reconsider the dynamic building of a request, and go for a simpler interface of the sessions class, similar to how it  was in in version 1. 

What's your take on this?

**Comment by niro1987** [comment]
You are right, in the current way of working, if one of the requests fails, the other(s) still proceed.
Home Assistant can simply create it's own overview, as can any other use-case.
